google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/project/ProjectSelectionDialog.java

```diff
@@ -33,6 +104,363 @@
   @Nullable
   CloudProject showDialog(@Nullable CloudProject cloudProject) {
-    return cloudProject;
+    // create dialog wrapper for the form once, create services and subscribe to login changes once.
+    if (dialogWrapper == null) {
+      dialogWrapper = new ProjectSelectionDialogWrapper();
+
+      dialogWrapper.setTitle(GctBundle.getString("cloud.project.selector.dialog.title"));
+      // disabled unless project is selected in the list.
+      dialogWrapper.setOKActionEnabled(false);
+      // install additional IDEA UI.
+      installTableSpeedSearch(projectListTable);
+
+      Stream.of(ProjectManager.getInstance().getOpenProjects())
+          .forEach(
+              project ->
+                  project
+                      .getMessageBus()
+                      .connect(dialogWrapper.getDisposable() /* disconnect once dialog is gone. */)
+                      .subscribe(
+                          GoogleLoginListener.GOOGLE_LOGIN_LISTENER_TOPIC, this::loadAllProjects));
+    }
+
+    loadAllProjects();
+    setSelectedProject(cloudProject);
+
+    DialogManager.show(dialogWrapper);
+
+    int result = dialogWrapper.getExitCode();
+    return result == DialogWrapper.OK_EXIT_CODE ? getSelectedProject() : null;
+  }
+
+  @VisibleForTesting
+  void setSelectedProject(CloudProject cloudProject) {
+    if (cloudProject != null && !Strings.isNullOrEmpty(cloudProject.googleUsername())) {
+      Optional<CredentialedUser> loggedInUser =
+          Services.getLoginService().getLoggedInUser(cloudProject.googleUsername());
+      if (loggedInUser.isPresent()) {
+        selectedProjectsByAccount.put(loggedInUser.get(), cloudProject.projectName());
+        accountComboBox.setSelectedItem(loggedInUser.get());
+      } else {
+        // specified user is not in logged in user list, clear the account selection.
+        accountComboBox.setSelectedItem(null);
+      }
+    } else {
+      // no project set, show active user projects by default.
+      accountComboBox.setSelectedItem(Services.getLoginService().getActiveUser());
+    }
+  }
+
+  @VisibleForTesting
+  CloudProject getSelectedProject() {
+    CredentialedUser user = accountComboBox.getItemAt(accountComboBox.getSelectedIndex());
+    return CloudProject.create(getSelectedProjectName(), getSelectedProjectId(), user.getEmail());
+  }
+
+  @VisibleForTesting
+  void loadAllProjects() {
+    Collection<CredentialedUser> credentialedUsers =
+        Services.getLoginService().getAllUsers().values();
+    if (credentialedUsers.isEmpty()) {
+      showSignInRequest();
+    } else {
+      hideSignInRequest();
+      cachedProjectList.clear();
+      accountComboBox.removeAllItems();
+      for (CredentialedUser user : credentialedUsers) {
+        accountComboBox.addItem(user);
+        loadProjectList(user);
+      }
+    }
+  }
+
+  private void loadProjectList(CredentialedUser user) {
+    // check if cached, if not then check if already loading, and finally start loading if not.
```

Comment:
could you promote top level comments like this that apply to the whole method to actually be javadoc method comments? I know some of these cases are for private methods, but I still think it makes sense.

Reply:
sure, great idea.